File: FacturaInvent2/Network/APIClient.swift
//
//  APIClient.swift
//  FacturaInvent2
//
//  Cliente HTTP base: URLSession + JWT + envelope { data, error }.
//

import Foundation
import Observation

enum APIError: LocalizedError {
    case servidorNoConfigurado
    case noAutenticado
    case servidor(String)
    case sinConexion(String)
    case respuestaInvalida

    var errorDescription: String? {
        switch self {
        case .servidorNoConfigurado:
            return "Configura la URL del servidor en Ajustes"
        case .noAutenticado:
            return "La sesión expiró, inicia sesión de nuevo"
        case .servidor(let mensaje):
            return mensaje
        case .sinConexion(let detalle):
            return "Sin conexión con el servidor: \(detalle)"
        case .respuestaInvalida:
            return "El servidor devolvió una respuesta inesperada"
        }
    }
}

/// Envelope estándar del backend: { "data": ..., "error": null }
private struct APIEnvelope<T: Decodable>: Decodable {
    let data: T?
    let error: String?
}

@Observable
final class APIClient {

    static let shared = APIClient()

    private(set) var token: String? {
        didSet { UserDefaults.standard.set(token, forKey: Self.claveToken) }
    }

    private(set) var usuarioActual: Usuario? {
        didSet {
            let data = usuarioActual.flatMap { try? JSONEncoder().encode($0) }
            UserDefaults.standard.set(data, forKey: Self.claveUsuario)
        }
    }

    var sesionActiva: Bool { token != nil }

    private static let claveToken = "api.token"
    private static let claveUsuario = "api.usuario"

    private let decoder: JSONDecoder
    private let encoder: JSONEncoder
    private let session: URLSession

    private init() {
        decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        encoder = JSONEncoder()
        encoder.keyEncodingStrategy = .convertToSnakeCase

        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 15
        session = URLSession(configuration: config)

        token = UserDefaults.standard.string(forKey: Self.claveToken)
        if let data = UserDefaults.standard.data(forKey: Self.claveUsuario) {
            usuarioActual = try? JSONDecoder().decode(Usuario.self, from: data)
        }
    }

    // MARK: - Sesión

    func login(username: String, password: String) async throws {
        let respuesta: LoginResponse = try await post(
            "auth/login",
            body: CredencialesLogin(username: username, password: password),
            autenticado: false
        )
        token = respuesta.token
        usuarioActual = respuesta.usuario
    }

    func logout() {
        token = nil
        usuarioActual = nil
    }

    // MARK: - Métodos HTTP

    func get<T: Decodable>(_ path: String, query: [URLQueryItem] = []) async throws -> T {
        try await request(method: "GET", path: path, query: query, body: Optional<Int>.none)
    }

    func post<T: Decodable, B: Encodable>(_ path: String, body: B, autenticado: Bool = true) async throws -> T {
        try await request(method: "POST", path: path, body: body, autenticado: autenticado)
    }

    func put<T: Decodable, B: Encodable>(_ path: String, body: B) async throws -> T {
        try await request(method: "PUT", path: path, body: body)
    }

    func delete(_ path: String) async throws {
        let _: Bool = try await request(method: "DELETE", path: path, body: Optional<Int>.none)
    }

    /// Descarga un archivo (ej. export .xlsx) a un archivo temporal y retorna su URL local.
    func descargarArchivo(_ path: String, query: [URLQueryItem] = [], nombre: String) async throws -> URL {
        let request = try construirRequest(method: "GET", path: path, query: query, bodyData: nil, autenticado: true)
        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse else { throw APIError.respuestaInvalida }
            guard http.statusCode == 200 else {
                throw APIError.servidor(mensajeDeError(en: data) ?? "Error \(http.statusCode)")
            }
            let destino = FileManager.default.temporaryDirectory.appendingPathComponent(nombre)
            try? FileManager.default.removeItem(at: destino)
            try data.write(to: destino)
            return destino
        } catch let error as APIError {
            throw error
        } catch {
            throw APIError.sinConexion(error.localizedDescription)
        }
    }

    // MARK: - Implementación

    private func request<T: Decodable, B: Encodable>(
        method: String,
        path: String,
        query: [URLQueryItem] = [],
        body: B?,
        autenticado: Bool = true
    ) async throws -> T {
        let bodyData = try body.map { try encoder.encode($0) }
        let urlRequest = try construirRequest(
            method: method, path: path, query: query, bodyData: bodyData, autenticado: autenticado
        )

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: urlRequest)
        } catch {
            throw APIError.sinConexion(error.localizedDescription)
        }

        guard let http = response as? HTTPURLResponse else { throw APIError.respuestaInvalida }

        if http.statusCode == 401 && autenticado {
            logout()
            throw APIError.noAutenticado
        }

        let envelope: APIEnvelope<T>
        do {
            envelope = try decoder.decode(APIEnvelope<T>.self, from: data)
        } catch {
            throw APIError.respuestaInvalida
        }

        if let error = envelope.error {
            throw APIError.servidor(error)
        }
        guard let resultado = envelope.data else { throw APIError.respuestaInvalida }
        return resultado
    }

    private func construirRequest(
        method: String,
        path: String,
        query: [URLQueryItem],
        bodyData: Data?,
        autenticado: Bool
    ) throws -> URLRequest {
        guard let base = ServerConfig.shared.baseURL else { throw APIError.servidorNoConfigurado }

        var components = URLComponents(
            url: base.appendingPathComponent(path), resolvingAgainstBaseURL: false
        )
        if !query.isEmpty { components?.queryItems = query }
        guard let url = components?.url else { throw APIError.servidorNoConfigurado }

        var request = URLRequest(url: url)
        request.httpMethod = method
        if let bodyData {
            request.httpBody = bodyData
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        if autenticado {
            guard let token else { throw APIError.noAutenticado }
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        return request
    }

    private func mensajeDeError(en data: Data) -> String? {
        struct SoloError: Decodable { let error: String? }
        return (try? decoder.decode(SoloError.self, from: data))?.error
    }
}

private struct CredencialesLogin: Encodable {
    let username: String
    let password: String
}

File: flutter_app/lib/api/api_client.dart
// Cliente HTTP base: JWT + envelope { data, error }.
// Equivalente a APIClient.swift de la app iOS.

import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';

import '../models/usuario.dart';
import 'server_config.dart';

class APIException implements Exception {
  APIException(this.mensaje, {this.sinConexion = false, this.noAutenticado = false});
  final String mensaje;
  final bool sinConexion;
  final bool noAutenticado;

  @override
  String toString() => mensaje;
}

class APIClient {
  APIClient._();
  static final APIClient shared = APIClient._();

  static const _claveToken = 'api.token';
  static const _claveUsuario = 'api.usuario';

  String? token;
  Usuario? usuarioActual;

  bool get sesionActiva => token != null;

  Future<void> cargar() async {
    final prefs = await SharedPreferences.getInstance();
    token = prefs.getString(_claveToken);
    final usuarioJson = prefs.getString(_claveUsuario);
    if (usuarioJson != null) {
      usuarioActual = Usuario.fromJson(jsonDecode(usuarioJson));
    }
  }

  Future<void> login(String username, String password) async {
    final data = await post(
      'auth/login',
      body: {'username': username, 'password': password},
      autenticado: false,
    );
    token = data['token'] as String;
    usuarioActual = Usuario.fromJson(data['usuario']);
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_claveToken, token!);
    await prefs.setString(_claveUsuario, jsonEncode(data['usuario']));
  }

  Future<void> logout() async {
    token = null;
    usuarioActual = null;
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_claveToken);
    await prefs.remove(_claveUsuario);
  }

  Future<dynamic> get(String path, {Map<String, String>? query}) =>
      _request('GET', path, query: query);

  Future<dynamic> post(String path, {Object? body, bool autenticado = true}) =>
      _request('POST', path, body: body, autenticado: autenticado);

  Future<dynamic> put(String path, {Object? body}) =>
      _request('PUT', path, body: body);

  Future<dynamic> delete(String path) => _request('DELETE', path);

  /// Descarga un archivo binario (ej. export .xlsx) y retorna los bytes.
  Future<List<int>> descargarArchivo(String path, {Map<String, String>? query}) async {
    final uri = _uri(path, query);
    final respuesta = await http.get(uri, headers: _headers(autenticado: true));
    if (respuesta.statusCode != 200) {
      throw APIException('Error ${respuesta.statusCode} al descargar');
    }
    return respuesta.bodyBytes;
  }

  Uri _uri(String path, Map<String, String>? query) {
    final base = ServerConfig.shared.baseUri;
    return base.replace(path: '/$path', queryParameters: query);
  }

  Map<String, String> _headers({required bool autenticado}) {
    final headers = {'Content-Type': 'application/json'};
    if (autenticado && token != null) {
      headers['Authorization'] = 'Bearer $token';
    }
    return headers;
  }

  Future<dynamic> _request(
    String method,
    String path, {
    Map<String, String>? query,
    Object? body,
    bool autenticado = true,
  }) async {
    if (autenticado && token == null) {
      throw APIException('La sesión expiró, inicia sesión de nuevo', noAutenticado: true);
    }

    final uri = _uri(path, query);
    final request = http.Request(method, uri);
    request.headers.addAll(_headers(autenticado: autenticado));
    if (body != null) {
      request.body = jsonEncode(body);
    }

    http.Response respuesta;
    try {
      final streamed = await request.send().timeout(const Duration(seconds: 15));
      respuesta = await http.Response.fromStream(streamed);
    } catch (e) {
      throw APIException('Sin conexión con el servidor: $e', sinConexion: true);
    }

    if (respuesta.statusCode == 401 && autenticado) {
      await logout();
      throw APIException('La sesión expiró, inicia sesión de nuevo', noAutenticado: true);
    }

    final Map<String, dynamic> envelope;
    try {
      envelope = jsonDecode(utf8.decode(respuesta.bodyBytes));
    } catch (_) {
      throw APIException('El servidor devolvió una respuesta inesperada');
    }

    if (envelope['error'] != null) {
      throw APIException(envelope['error'] as String);
    }
    return envelope['data'];
  }
}
